File: 1010/hashmap.cpp
#include <gtest/gtest.h>

#include <cstdlib>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  int numPairsDivisibleBy60(vector<int>& time) {
    unordered_map<int, size_t> counts;
    for (const int& t : time) {
      counts[t % 60] += 1;
    }
    int num = 0;
    for (int i = 1; i < 30; ++i) {
      if (counts.find(i) != counts.end() && counts.find(60-i) != counts.end()) {
        num += counts[i] * counts[60-i];
      }
    }
    if (counts.find(0) != counts.end()) {
      num += counts[0] * (counts[0] - 1) / 2;
    }
    if (counts.find(30) != counts.end()) {
      num += counts[30] * (counts[30] - 1) / 2;
    }
    return num;
  }
};

int main() {
  Solution sol;
  vector<int> times;
  int expected;

  times = {30, 20, 150, 100, 40};
  expected = 3;
  EXPECT_EQ(expected, sol.numPairsDivisibleBy60(times));

  times = {60, 60, 60};
  expected = 3;
  EXPECT_EQ(expected, sol.numPairsDivisibleBy60(times));

  exit(EXIT_SUCCESS);
}

File: 1010/hashmap.go
package main

import (
	"fmt"
	"os"
)

func numPairsDivisibleBy60(time []int) int {
	counts := make(map[int]int)
	ret := 0
	for _, t := range time {
		remainder := t % 60
		v, b := counts[(60-remainder)%60]
		if b {
			ret += v
		}
		counts[remainder] += 1
	}
	return ret
}

func main() {
	var time []int
	var actual, expected int

	time, expected = []int{30, 20, 150, 100, 40}, 3
	actual = numPairsDivisibleBy60(time)
	if actual != expected {
		fmt.Fprintln(os.Stderr, "time:", time, "actual:", actual, "expected:", expected)
	}

	time, expected = []int{60, 60, 60}, 3
	actual = numPairsDivisibleBy60(time)
	if actual != expected {
		fmt.Fprintln(os.Stderr, "time:", time, "actual:", actual, "expected:", expected)
	}
}
